server/session/handler_item_stack_request.go

```diff
@@ -479,6 +483,190 @@ func (h *ItemStackRequestHandler) handleMineBlock(a *protocol.MineBlockStackRequ
+	// First check if there actually is an anvil opened.
+	if !s.containerOpened.Load() {
+		return fmt.Errorf("no anvil container opened")
+	}
+	pos := s.openedPos.Load()
+	w := s.c.World()
+	anvil, ok := w.Block(pos).(block.Anvil)
+	if !ok {
+		return fmt.Errorf("no anvil container opened")
+	}
+	if len(filterStrings) < int(a.FilterStringIndex) {
+		// Invalid filter string index.
+		return nil
+	}
+
+	first, _ := h.itemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerAnvilInput,
+		Slot:        1,
+	}, s)
+	if first.Empty() {
+		// First anvil slot is empty, can't result in anything.
+		return nil
+	}
+	result := first
+
+	second, _ := h.itemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID: containerAnvilMaterial,
+		Slot:        0x2,
+	}, s)
+
+	var cost, repairCount int
+	var resultEnchantments []item.Enchantment
+	if !second.Empty() {
+		if repairable, ok := first.Item().(item.Repairable); ok && repairable.RepairableBy(second) {
+			d := min(first.MaxDurability()-first.Durability(), first.MaxDurability()/4)
+			if d <= 0 {
+				return nil
+			}
+
+			for ; d > 0 && repairCount < second.Count(); repairCount, d = repairCount+1, min(result.MaxDurability()-result.Durability(), result.MaxDurability()/4) {
+				result = result.WithDurability(result.Durability() + d)
+				cost++
+			}
+		} else {
+			_, ok := second.Item().(item.EnchantedBook)
+			enchant := ok && len(second.Enchantments()) > 0
+			_, durable := first.Item().(item.Durable)
+			if !enchant && (first.Item() != second.Item() || !durable) {
+				return nil
+			}
+			if durable && !enchant {
+				d := first.MaxDurability() - (first.Durability() + (second.Durability() + first.MaxDurability()*12/100))
+				if d < 0 {
+					d = 0
+				}
+				if d < first.MaxDurability()-first.Durability() {
+					result = result.WithDurability(d)
+					cost += 2
+				}
+			}
+
+			for _, e := range second.Enchantments() {
+				t := e.Type()
+				firstLevel := 0
+				if firstEnchant, ok := first.Enchantment(t); ok {
+					firstLevel = firstEnchant.Level()
+				}
+				resultLevel := max(firstLevel, e.Level())
+				if firstLevel == e.Level() {
+					resultLevel = firstLevel + 1
+				}
+				compatible := t.CompatibleWithItem(first.Item())
+				if _, ok := first.Item().(item.EnchantedBook); ok {
+					compatible = true
+				}
+				for _, e2 := range first.Enchantments() {
+					if t != e2.Type() && !t.CompatibleWithOther(e2.Type()) {
+						compatible = false
+						cost++
+					}
+				}
+				if compatible {
+					if resultLevel > t.MaxLevel() {
+						resultLevel = t.MaxLevel()
+					}
+					resultEnchantments = append(resultEnchantments, item.NewEnchantment(t, resultLevel))
+					rarityCost := t.Rarity().ApplyCost
+					if enchant {
+						rarityCost = max(1, rarityCost/2)
+					}
+					cost += rarityCost * resultLevel
+					if first.Count() > 1 {
+						cost = 40
+					}
+				} else {
+					return nil
```

Comment:
This doesn't seem quite right. Anvil use should only fail if both **a)** not repairing, and **b)** no enchantments from the second item are compatible

Reply:
This condition should never be met if repairing though?

Reply:
The second point is valid though

Reply:
> This condition should never be met if repairing though?
It can if repairing with another damaged item

Reply:
I think this is resolved with the latest commit, let me know if not.